category_theory/limits/binary_products.lean

```diff
@@ -0,0 +1,490 @@
+-- Copyright (c) 2018 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison, Reid Barton, Mario Carneiro
+
+import category_theory.limits.shape
+
+open category_theory
+
+namespace category_theory.limits
+
+universes u v w
+
+variables {C : Type u} [𝒞 : category.{u v} C]
+include 𝒞
+
+section binary_product
+structure is_binary_product {Y Z : C} (t : span Y Z) :=
+(lift : ∀ (s : span Y Z), s.X ⟶ t.X)
+(fac₁' : ∀ (s : span Y Z), (lift s) ≫ t.π₁ = s.π₁ . obviously)
+(fac₂' : ∀ (s : span Y Z), (lift s) ≫ t.π₂ = s.π₂ . obviously)
+(uniq' : ∀ (s : span Y Z) (m : s.X ⟶ t.X) (w₁ : m ≫ t.π₁ = s.π₁) (w₂ : m ≫ t.π₂ = s.π₂), m = lift s . obviously)
+
+restate_axiom is_binary_product.fac₁'
+attribute [simp] is_binary_product.fac₁
+restate_axiom is_binary_product.fac₂'
+attribute [simp] is_binary_product.fac₂
+restate_axiom is_binary_product.uniq'
+
+@[extensionality] lemma is_binary_product.ext {Y Z : C} {t : span Y Z} (P Q : is_binary_product t) : P = Q :=
+begin
+  tactic.unfreeze_local_instances,
+  cases P, cases Q,
+  congr,
+  ext1,
+  exact eq.symm (P_uniq' x (Q_lift x) (Q_fac₁' x) (Q_fac₂' x))
+end
+
+instance subsingleton_is_binary_product {Y Z : C} {t : span Y Z} : subsingleton (is_binary_product t) := by obviously
+
+lemma is_binary_product.uniq'' {Y Z : C} {t : span Y Z} (h : is_binary_product t) {X' : C} (m : X' ⟶ t.X) :
```

Comment:
where is the `uniq` without prime? Does it make sense to remove the prime here?

Reply:
These lemmas are pretty reasonable as `hom_desc` and `hom_lift`. (I'm right that `hom_desc` is better than `hom_eq_desc` if the point is to say "a hom is equal to a descent"?)